onnxruntime/python/tools/quantization/execution_providers/qnn/quant_config.py

```diff
@@ -0,0 +1,79 @@
+# license information.
+# --------------------------------------------------------------------------
+from pathlib import Path
+
+import onnx
+
+from ...calibrate import CalibrationDataReader, CalibrationMethod
+from ...quant_utils import QuantType
+from ...quantize import StaticQuantConfig
+
+Q16_TYPES = {QuantType.QInt16, QuantType.QUInt16}
+Q8_TYPES = {QuantType.QInt8, QuantType.QUInt8}
+OP_TYPES_TO_EXCLUDE = {"Cast", "Transpose"}
+
+
+def get_qnn_qdq_config(
+    model_input: Path,
+    calibration_data_reader: CalibrationDataReader,
+    calibrate_method=CalibrationMethod.MinMax,
+    activation_type=QuantType.QUInt8,
+    weight_type=QuantType.QUInt8,
+):
+    # Parse model nodes to setup overrides.
+    model = onnx.load_model(model_input)
+
+    op_types = set()
+    tensor_quant_overrides = {}
+
+    name_to_initializer = {initializer.name: initializer for initializer in model.graph.initializer}
+
+    for node in model.graph.node:
+        op_types.add(node.op_type)
+
+        if node.op_type == "MatMul" and activation_type in Q16_TYPES and weight_type in Q8_TYPES:
+            weight_symmetric = weight_type == QuantType.QInt8
+
+            # Override initializers to use the weight_type
+            for input_name in node.input:
+                if input_name in name_to_initializer:
+                    tensor_quant_overrides[input_name] = {"quant_type": weight_type, "symmetric": weight_symmetric}
+        elif node.op_type == "LayerNormalization" and activation_type in Q16_TYPES and weight_type in Q8_TYPES:
+            weight_symmetric = weight_type == QuantType.QInt8
+
+            # Override initializers to use the weight_type. Don't override the bias input.
+            for i in range(2):
+                input_name = node.input[i]
+                if input_name in name_to_initializer:
+                    tensor_quant_overrides[input_name] = {"quant_type": weight_type, "symmetric": weight_symmetric}
+        elif node.op_type == "Sigmoid":
+            if activation_type == QuantType.QUInt16:
+                tensor_quant_overrides[node.output[0]] = {"scale": 1.0 / 65536.0, "zero_point": 0}
+            elif activation_type == QuantType.QInt16:
+                tensor_quant_overrides[node.output[0]] = {"scale": 1.0 / 32768.0, "zero_point": 0}
+        elif node.op_type == "Tanh":
+            if activation_type == QuantType.QUInt16:
+                tensor_quant_overrides[node.output[0]] = {"scale": 1.0 / 32768.0, "zero_point": 32768}
+            elif activation_type == QuantType.QInt16:
+                tensor_quant_overrides[node.output[0]] = {"scale": 1.0 / 32768.0, "zero_point": 0}
+
+    extra_options = {
+        "MinimumRealRange": 0.0001,
+        "DedicatedQDQPair": False,  # Let ORT optimizer duplicate DQ nodes
+        "TensorQuantOverrides": tensor_quant_overrides,
+    }
+
+    if activation_type in Q16_TYPES or weight_type in Q16_TYPES:
+        extra_options["UseQDQContribOps"] = True
```

Comment:
>extra_options["UseQDQContribOps"] = True [](start = 8, length = 40)

Do we still need this since 16bit support was merged to Onnx already?
Or do we need to wait for Ort to upgrade Onnx submodule?

Reply:
We need to wait for the next ONNX release and then the next ORT release that contains the new ONNX version. I'll leave a TODO to change this once ONNX version 21 is available.

Reply:
next ONNX release won't be until ~Feb